add groups

KateF20 opened the pull request:
create group template and path to it
show the button to create group on the top of the page (added to main/base.html) 
(only) started working with forms and models, a new form for groups added 
create group method for views

это прям самое самое начало

KateF20 committed:
create group template and path to it
show the button to create group on the top of the page (added to main/base.html)
started working with forms and models, a new form for groups added
create group method for views

KateF20 committed:
Group model registered for admin site
header naming changes
group assignment added to when creating task
group filter added to the main page
verbose name added to Group model
groups added to the main page

homm commented on taskmanager/taskeditor/forms.py:
```diff
@@ -32,3 +32,19 @@ class NewUserForm(UserCreationForm):
+
+    def save(self, commit=True, user=None):
+        self.instance.user = user
+        return super(GroupForm, self).save(commit=commit)
```
И выше. Может ещё где-то

```suggestion
        return super().save(commit=commit)
```

homm commented on taskmanager/taskeditor/models.py:
```diff
@@ -2,11 +2,24 @@
 
+class Group(models.Model):
+    group = models.CharField('Group', max_length=50)
+    user = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, blank=True)
```
```suggestion
    user = models.ForeignKey(User, on_delete=models.CASCADE)
```

homm commented on taskmanager/taskeditor/models.py:
```diff
@@ -2,11 +2,24 @@
 
 
+class Group(models.Model):
+    group = models.CharField('Group', max_length=50)
```
```suggestion
    name = models.CharField('Group', max_length=50)
```

homm commented on taskmanager/taskeditor/templates/main/create.html:
```diff
@@ -10,7 +10,8 @@ <h1 style="color: royalblue">Create task</h1>
         {{ form.title }} <br>
         {{ form.task }} <br>
-        <button type="submit" class="btn btn-success" style="color: antiquewhite; background-color: royalblue">Add</button>
+        {{ form.group }} <br>
```
```suggestion
        <div>
        {{ form.group }} <br>
        {{ form.group.errors }}
        </div>
```

homm commented on taskmanager/taskeditor/views.py:
```diff
@@ -75,3 +78,20 @@ def registration(request):
+
+    context = {
+        'form': form,
+        'error': error
```
Не используется

KateF20 committed:
renamed Group.group --> Group.name
groups can be deleted (changed index.html, urls and views)
minor changes in delete_task.html
super() method changed to a less explicit writing
filter by group id

KateF20 committed:
migrations for adding group: new group model, new field for task model

homm commented on taskmanager/taskeditor/models.py:
```diff
@@ -2,11 +2,24 @@
 
 
+class Group(models.Model):
+    name = models.CharField('Group', max_length=50)
```
```suggestion
    name = models.CharField('Name', max_length=50)
```

homm commented on taskmanager/taskeditor/templates/main/create.html:
```diff
@@ -10,7 +10,11 @@ <h1 style="color: royalblue">Create task</h1>
         {% csrf_token %}
         {{ form.title }} <br>
         {{ form.task }} <br>
```
Тут тоже нужно выводить errors

homm commented on taskmanager/taskeditor/templates/main/create_group.html:
```diff
@@ -0,0 +1,14 @@
+    <h1 style="color: royalblue">Create group</h1>
+    <form method="post">
+        {% csrf_token %}
+        {{ form.name }} <br>
```
errors

homm commented on taskmanager/taskeditor/templates/main/delete_task.html:
```diff
@@ -6,9 +6,12 @@
     {% csrf_token %}
-    <input type="submit" value="Delete">
+    <input class="btn btn-outline-secondary" type="button" type="submit" value="Delete">
+    <button class="btn btn-outline-secondary" type="button">
```
```suggestion
    <a href="{% url 'home' %}" class="btn btn-outline-secondary">
```

homm commented on taskmanager/taskeditor/templates/main/index.html:
```diff
@@ -22,12 +21,25 @@ <h2>Your task manager</h2>
+                  <li><hr class="dropdown-divider"></li>
+                   {% for elem in groups %}
+                        <li><a class="dropdown-item rounded-2" href="{% url 'home' status=elem.id%}">{{ elem.name}}</a></li>
+                       <button class="btn btn-outline-secondary" type="button">
```
Можно вкладывать только ul → li, другие элементы нельзя

```suggestion
                        <li><a class="dropdown-item rounded-2" href="{% url 'home' status=elem.id%}">{{ elem.name}}</a>
                       <a href="{% url 'delete_group' elem.id %}" class="btn btn-outline-secondary" type="button">
```

homm commented on taskmanager/taskeditor/templates/main/index.html:
```diff
@@ -22,12 +21,25 @@ <h2>Your task manager</h2>
+               <ul class="dropdown-menu position-static d-grid gap-1 p-2 rounded-3 mx-0 shadow w-220px" data-bs-theme="light">
+                  <li><a class="dropdown-item rounded-2" href="#">Filter groups</a></li>
+                  <li><hr class="dropdown-divider"></li>
+                   {% for elem in groups %}
```
```suggestion
                   {% for group in groups %}
```

KateF20 committed:
changed group and completion filters

KateF20 committed:
regex adjusted to another parameter
added current status and group-id for correct rendering of the main page

KateF20 committed:
buttons corrected

homm commented on taskmanager/taskeditor/views.py:
```diff
@@ -1,13 +1,14 @@
-def index(request, status=None):
+def index(request, status=None, group_id=None):
     tasks = []
+    groups = Group.objects.filter(user=request.user)
```
Что если пользователь не авторизован?

homm commented on taskmanager/taskeditor/urls.py:
```diff
@@ -2,8 +2,10 @@
 
 urlpatterns = [
-    re_path(r'^(?P<status>all|completed)?$', views.index, name='home'),
+    re_path(r'^(?P<group_id>[0-9]+)?\/?(?P<status>all|completed)?$', views.index, name='home'),
```
```suggestion
    re_path(
        r'^(?P<status>all|completed)?$',
        views.index, name='home'
    ),
    re_path(
        r'^(?P<group_id>[0-9]+)(?:/(?P<status>all|completed))?$',
        views.index, name='home'
    ),
```

KateF20 committed:
added custom tag url2

KateF20 committed:
minor changes in naming and rendering

homm commented on taskmanager/taskeditor/migrations/0006_rename_group_group_name.py:
Этот файл можно совместить в 0005.

* ./manage.py migrate taskeditor 0004
* rm taskmanager/taskeditor/migrations/{0005,0006}*
* ./manage.py makemigrations taskeditor
* ./manage.py migrate taskeditor

Ещё можно migrate запускать с `--fake`, тогда данные не потеряются
https://docs.djangoproject.com/en/4.2/ref/django-admin/#django-admin-migrate

homm commented on taskmanager/taskeditor/views.py:
```diff
@@ -8,8 +8,8 @@
     tasks = []
-    groups = Group.objects.filter(user=request.user)
     if request.user.is_authenticated:
+        groups = Group.objects.filter(user=request.user)
```
вопрос остается тот же, Что если пользователь не авторизован?

KateF20 committed:
minor changes for this branch, + migrations
